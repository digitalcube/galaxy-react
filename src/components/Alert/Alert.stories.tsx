import React from 'react';
import { Meta, Story } from '@storybook/react';
import { Alert, AlertProps } from './Alert';

const meta: Meta = {
  title: 'Components/Alert',
  component: Alert,
  argTypes: {
    type: {
      options: ['info', 'warning', 'danger', 'success'],
      control: { type: 'radio' },
    },
  },
  parameters: {
    controls: { expanded: true },
  },
};

export default meta;

const Template: Story<AlertProps> = (args) => <Alert {...args} />;

// By passing using the Args format for exported stories, you can control the props for a component for reuse in a test
// https://storybook.js.org/docs/react/workflows/unit-testing
export const Default = Template.bind({});
Default.args = {
  type: 'info',
  title: 'title',
  children: 'message',
  showIcon: true,
  className: '',
};
