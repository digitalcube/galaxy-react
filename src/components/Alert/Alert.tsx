import React, { FC, PropsWithChildren, ReactNode } from 'react';
import { useMemo } from 'react';
import { FaExclamationTriangle, FaExclamationCircle } from 'react-icons/fa';

export type AlertType = 'info' | 'warning' | 'success' | 'danger';
export type AlertProps = PropsWithChildren<{
  type: AlertType;
  showIcon?: boolean;
  icon?: ReactNode;
  title?: string;
  className?: string;
}>;

const AlertIcon: FC<Pick<AlertProps, 'type' | 'icon'>> = ({ type, icon }) => {
  if (icon) return <>{icon}</>;
  switch (type) {
    case 'warning':
      return <FaExclamationTriangle className="d-block" />;
    case 'danger':
    case 'success':
    case 'info':
    default:
      return <FaExclamationCircle className="d-block" />;
  }
};

export const Alert: FC<AlertProps> = ({
  children,
  type = 'info',
  icon,
  showIcon = true,
  title,
  className,
}) => {
  const classNames = useMemo(() => {
    const items = [
      `alert ${`alert-${type}`} mb-4 p-0 border-0 d-flex c-alert`,
      className,
    ];
    return items.filter(Boolean);
  }, [type, className]);
  if (!title && !children) return null;
  return (
    <div className={classNames.join(' ')} role="alert">
      {showIcon ? (
        <div className="alert-icon px-1 d-flex align-items-center">
          <i aria-hidden="true">
            <AlertIcon type={type} icon={icon} />
          </i>
        </div>
      ) : null}
      <div className="alert-body px-3 py-2 flex-grow-1">
        {title ? <h4 className="alert-heading">{title}</h4> : null}
        {typeof children === 'string' ? (
          <p className="mb-0">{children}</p>
        ) : (
          children
        )}
      </div>
    </div>
  );
};
