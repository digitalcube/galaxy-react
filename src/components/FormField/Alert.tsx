import React, { FC, PropsWithChildren } from 'react';
import { Alert } from '../Alert/Alert';

export type FormAlertProps = PropsWithChildren<{
  errorMessage?: string;
}>;
export const FormAlert: FC<FormAlertProps> = ({ errorMessage, children }) => {
  if (!errorMessage && !children) return null;
  return (
    <Alert type="danger">
      <p className="mb-0">{errorMessage}</p>
      {children}
    </Alert>
  );
};
