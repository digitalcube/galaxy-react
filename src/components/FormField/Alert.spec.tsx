import { render } from '@testing-library/react';
import React from 'react';
import { Default as Alert } from './Alert.stories';

describe('Alert', () => {
  it('to match snapshot [no item]', () => {
    const { asFragment } = render(<Alert {...(Alert.args as any)} />);
    expect(asFragment()).toMatchSnapshot();
  });
  it('to match snapshot [error message]', () => {
    const { asFragment } = render(<Alert errorMessage="error" />);
    expect(asFragment()).toMatchSnapshot();
  });
  it('to match snapshot [children]', () => {
    const { asFragment } = render(<Alert>children</Alert>);
    expect(asFragment()).toMatchSnapshot();
  });
  it('to match snapshot [mixed]', () => {
    const { asFragment } = render(
      <Alert errorMessage="message">children</Alert>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
