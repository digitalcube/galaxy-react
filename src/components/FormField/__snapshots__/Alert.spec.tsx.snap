// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Alert to match snapshot [children] 1`] = `
<DocumentFragment>
  <div
    class="mb-4 p-0 border-0 d-flex c-alert alert alert-danger fade"
    role="alert"
  >
    <div
      class="alert-icon px-1 d-flex align-items-center"
    >
      <i
        aria-hidden="true"
      >
        <svg
          class="d-block"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 512 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M504 256c0 136.997-111.043 248-248 248S8 392.997 8 256C8 119.083 119.043 8 256 8s248 111.083 248 248zm-248 50c-25.405 0-46 20.595-46 46s20.595 46 46 46 46-20.595 46-46-20.595-46-46-46zm-43.673-165.346l7.418 136c.347 6.364 5.609 11.346 11.982 11.346h48.546c6.373 0 11.635-4.982 11.982-11.346l7.418-136c.375-6.874-5.098-12.654-11.982-12.654h-63.383c-6.884 0-12.356 5.78-11.981 12.654z"
          />
        </svg>
      </i>
    </div>
    <div
      class="alert-body px-3 py-2 flex-grow-1"
    >
      <p
        class="mb-0"
      />
      children
    </div>
  </div>
</DocumentFragment>
`;

exports[`Alert to match snapshot [error message] 1`] = `
<DocumentFragment>
  <div
    class="mb-4 p-0 border-0 d-flex c-alert alert alert-danger fade"
    role="alert"
  >
    <div
      class="alert-icon px-1 d-flex align-items-center"
    >
      <i
        aria-hidden="true"
      >
        <svg
          class="d-block"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 512 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M504 256c0 136.997-111.043 248-248 248S8 392.997 8 256C8 119.083 119.043 8 256 8s248 111.083 248 248zm-248 50c-25.405 0-46 20.595-46 46s20.595 46 46 46 46-20.595 46-46-20.595-46-46-46zm-43.673-165.346l7.418 136c.347 6.364 5.609 11.346 11.982 11.346h48.546c6.373 0 11.635-4.982 11.982-11.346l7.418-136c.375-6.874-5.098-12.654-11.982-12.654h-63.383c-6.884 0-12.356 5.78-11.981 12.654z"
          />
        </svg>
      </i>
    </div>
    <div
      class="alert-body px-3 py-2 flex-grow-1"
    >
      <p
        class="mb-0"
      >
        error
      </p>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Alert to match snapshot [mixed] 1`] = `
<DocumentFragment>
  <div
    class="mb-4 p-0 border-0 d-flex c-alert alert alert-danger fade"
    role="alert"
  >
    <div
      class="alert-icon px-1 d-flex align-items-center"
    >
      <i
        aria-hidden="true"
      >
        <svg
          class="d-block"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 512 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M504 256c0 136.997-111.043 248-248 248S8 392.997 8 256C8 119.083 119.043 8 256 8s248 111.083 248 248zm-248 50c-25.405 0-46 20.595-46 46s20.595 46 46 46 46-20.595 46-46-20.595-46-46-46zm-43.673-165.346l7.418 136c.347 6.364 5.609 11.346 11.982 11.346h48.546c6.373 0 11.635-4.982 11.982-11.346l7.418-136c.375-6.874-5.098-12.654-11.982-12.654h-63.383c-6.884 0-12.356 5.78-11.981 12.654z"
          />
        </svg>
      </i>
    </div>
    <div
      class="alert-body px-3 py-2 flex-grow-1"
    >
      <p
        class="mb-0"
      >
        message
      </p>
      children
    </div>
  </div>
</DocumentFragment>
`;

exports[`Alert to match snapshot [no item] 1`] = `<DocumentFragment />`;
